### PR TITLE
Create bigger rooms with a single click

### PR DESCRIPTION
**Pull request by Loobinex**
When placing a room type, the player can use LSHIFT, LCTRL, and the NUMKEYS to place multiple room slabs with a single click. The bounding box displays this multi-slab room, to show the player the slabs that the current room type will be placed in.

These new features are described in the following sections:
- [Room Detection mode: Overview](https://github.com/Loobinex/keeperfx-unofficial/pull/209#issuecomment-698643521)
- [Fixed Room Size mode: Overview](https://github.com/Loobinex/keeperfx-unofficial/pull/209#issuecomment-699642228)
- [Bridge Painting mode: Overview](https://github.com/Loobinex/keeperfx-unofficial/pull/209#issuecomment-699660848)
- [Room Detection mode: Brief Technical Description](https://github.com/Loobinex/keeperfx-unofficial/pull/209#issuecomment-698652636)

Summary of code changes in this Pull Request:

_Code additions for the "find best room algorithm" (list to be completed):_
- Created **slab_room_detection.c** and **slab_room_detection.h**. These contain the algorithm for finding the "best" room, based on the current cursor position.
- ...

_Function changes/additions to support "box rooms" and "fancy rooms":_
- **_lots of stuff in slab_data.c_**
- **render_room** is used publicly to enable the drawing of a "fancy" bounding box. A Fancy bounding box is only drawn around the slabs that are in the room, and can be any shape.
- **create_map_volume_box()** and **create_frontview_map_volume_box()** updated to work with varying sizes of boxes.
- **create_accurate_map_volume_box()** and **create_accurate_frontview_map_volume_box()** created to allow for drawing "fancy" bounding boxes.
- **draw_frontview_engine()** and **draw_view()** now detect if the current room is a fancy room or a box room, and call the correct volume box function accordingly.
- **tag_cursor_blocks_place_room()** can now use a room of a chosen width/height, or a "fancy" room, and set up the bounding box correctly.
- **tag_cursor_blocks_sell_area()** can now use a room of a chosen width/height, and set up the bounding box correctly.
- **process_dungeon_control_packet_dungeon_build_room()** now sets the width/height of a "room" based on what input the player uses. **get_biggest_room()** is called from here, which runs the "find best room algorithm".
- **process_dungeon_control_packet_sell_operation()** now sets the width/height of a "room" based on what input the player uses.
- **gui_area_big_room_button()** uses the number of slabs in a room to display the correct total price.
- **player_build_room_at()** now uses the total room cost, rather the cost of a single slab.

_Additional supporting function changes:_
- **floor_height_for_volume_box()** moved from thing_traps.c to engine_render.c, so that it can be used by other scripts. Duplicate code was removed from tag_cursor_blocks_sell_area(), tag_cursor_blocks_place_door(), and tag_cursor_blocks_place_room() in main.cpp and replaced with a call to floor_height_for_volume_box().
- **create_line_element()** now contains a check for space in the poly pool, this is so the check doesn't have to be done outside of this function.

_DK variable name changes:_
- **field_2** of struct PlayerInfo renamed to **boxsize**. This is the number of slabs in a room/box.
- **temp_cluedo_mode** described (should maybe be renamed?) - 4891d95.
- **field_13** of struct MapVolumeBox renamed to **floor_height_z** this stores the height of a slab's lowest point. floor_height_for_volume_box() is used to set a height of 0 for water/lava, 1 for floors, and 5 for walls. If cluedo mode is used a wall will have a height of 2.
- Can **map_volume_box.field_17** be renamed too? To something like box_length, box_diameter, box_longest_side etc

### CODE REVIEW

**Comment by eddebaby** [comment]
This follows from #161

**Comment by eddebaby** [comment]
_**Room detection mode:** Overview_
image
- A "room" of a defined width and height can be detected/chosen when placing any room type, except bridges. 
- Its boundaries are indicated by an appropriate bounding box around the cursor.
- When the player clicks, the chosen room type is placed on all of the indicated tiles.
- If the player doesn't have enough funds to place all of the slabs, the bounding box is red, and the slabs will not be placed when the player clicks.

**Comment by eddebaby** [comment]
_**Room detection mode:** Brief Technical Description_

Rooms can either be:
- A "box" - i.e. they are square/rectangular, and all of the tiles within the rooms area are valid locations for the selected room type.
- A "fancy room" - this is any room that contains invalid locations for the selected room type. Fancy rooms are expressed as a 2D array of booleans, where a value of true means that the current tile is a valid location for the selected room type, and false means the opposite.

Drawing bounding boxes:
- create_map_volume_box() / create_frontview_map_volume_box() have been expanded, and are used to draw the bound boxes for _box rooms_. They use the width and height of the room.
- create_accurate_map_volume_box() / create_accurate_frontview_map_volume_box()  are new, and are used to draw the bound boxes for _fancy rooms_. They use the 2D array that describes the room. The correct edges are drawn, based on the presence of surrounding false ("air") tiles, so that the accurate outline of the room is displayed. 

Finding the biggest room:
- Starting at the current cursor position, and working outwards from there, each tile is considered as the centre of a potential room/
    - For each room centre, the biggest possible room (largest area) is found (by testing the validity of the tiles for placing the current room type).
    - This cursor must fall in the room, this ensures it is a room that we want, rather than a seemingly random one.
    - If a corridor (a room that is too wide) is found, then the biggest square room is returned instead.
- This is repeated 5 times, each time search for wider/taller rooms, these rooms are composited together to produce the final room.
- A composite room will ignore any found corridors, and if a composite room is found, it will take precedence over any "corridor room".

**Comment by eddebaby** [comment]
_**Fixed Room Size mode:** Overview_ 
image
image

- Hold LEFT CONTROL on the keyboard, after selecting a room type to place, to use Fixed Room Size mode.
- By default a 5x5 room, centred at the cursor, will be used. If the player scrolls the mouse wheel up and down, they can increase/decrease the size of the square room between 1x1 and 9x9 tiles.
- The player can also hold a numkey between 2 and 9 to select a square room of that size - e.g. hold numkey 5 for a 5x5 room.
- Only the tiles that are able to have the chosen room type placed on them will be surrounded by a green bounding box.
- When the player clicks, the chosen room type is placed on all of the indicated tiles.
- If the player doesn't have enough funds to place all of the slabs, the bounding box is red, and the slabs will not be placed when the player clicks.
- Can be used with any room type, except for Bridges.

**Comment by eddebaby** [comment]
_**Bridge Painting mode:** Overview_
image
- Hold LEFT CONTROL or LEFT SHIFT on the keyboard, after selecting the Bridge room type, to use Bridge Painting mode.
- Hold the left mouse button to continuously place Bridges. Move the mouse cursor around to "paint" a long bridge.

File: src/engine_render.c [hunk 1]
```diff
@@ -6027,73 +6211,197 @@ static void update_frontview_pointed_block(unsigned long laaa, unsigned char qdr
...
+    create_line_element(coord_x + box_width, coord_y,                      coord_x + box_width, coord_y + box_height,         coord_z - box_height,             map_volume_box.color);
+    create_line_element(coord_x,             coord_y + breadth,            coord_x,             coord_y + box_height + depth, coord_z - box_height + stl_width, map_volume_box.color);
+    create_line_element(coord_x + box_width, coord_y + breadth,            coord_x + box_width, coord_y + box_height + depth, coord_z - box_height + stl_width, map_volume_box.color);
+}
+void create_accurate_frontview_map_volume_box(struct RoomMap room_map, struct Camera *cam, unsigned char stl_width)
+{
+    unsigned char orient = ((unsigned int)(cam->orient_a + LbFPMath_PI/4) >> 9) & 0x03;
+    int floor_height_z = (map_volume_box.floor_height_z == 0) ? 1 : map_volume_box.floor_height_z; // ignore "liquid height", and force it to "floor height". All fancy rooms are on the ground, and this ensures the boundboxes are drawn correctly. A different solutino will be required if this function is used to draw fancy rooms over "liquid".
```
Comment (Loobinex) [inline src/engine_render.c:6272]: solutino is not a word

Comment (eddebaby) [reply src/engine_render.c:6272]: I think it is the Italian spelling ;p (Thank you I've corrected it)

**Comment by eddebaby** [comment]
Reorganisation:
- new roomspace scripts(roomspace.c/.h and roomspace_detection.c/.h) house the new functionality for dealing with roomspaces.
- engine_render.c contains the new fancy boundboxes, and improvements to the boundbox rendering in frontview
- All other scripts (apart from the new roomspace ones) either now call functions from the roomspace scripts, or have been tidied up

**Comment by Loobinex** [comment]
1) Boundbox colors are not right for selling. It's always colored for the slab beneath the cursor, so it could be a red boundbox but you still sell when clicking if there's other tiles below.
2) Ctrl+Scroll still both changes the size of the boundbox AND scrolls the objective message at the same time.

**Pull request merged**